google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/read_rows/StateMachine.java

```diff
@@ -0,0 +1,433 @@
+ * <p>Expected Usage:
+ *
+ * <pre>{@code
+ * StateMachine stateMachine = new StateMachine<>(myRowAdapter);
+ * while(responseIterator.hasNext()) {
+ *   ReadRowsResponse response = responseIterator.next();
+ *   if (!response.getLastScannedRowKey().isEmpty()) {
+ *     stateMachine.handleLastScannedRow(response.getLastScannedRowKey());
+ *   }
+ *
+ *   for(CellChunk chunk : response.getChunksList()) {
+ *     boolean hasCompleteRow = stateMachine.handleChunk(chunk);
+ *     if (hasCompleteRow) {
+ *       MyRow row = stateMachine.consumeRow();
+ *
+ *       if (myRowAdapter.isScanMarkerRow(row) {
+ *         // do something with marker
+ *       } else {
+ *         // do something with row
+ *       }
+ *     }
+ *   }
+ * }
+ * }</pre>
+ *
+ * <p>Package-private for internal use.
+ *
+ * @param <RowT> The type of row the adapter will build
+ */
+final class StateMachine<RowT> {
+  private final RowBuilder<RowT> adapter;
+  private State currentState;
+  private ByteString lastCompleteRowKey;
+
+  // Track current cell attributes: protocol omits them when they are repeated
+  private ByteString rowKey;
+  private String familyName;
+  private ByteString qualifier;
+  private long timestamp;
+  private List<String> labels;
+  private long expectedCellSize;
+  private long remainingCellBytes;
+
+  private RowT completeRow;
+
+  /**
+   * Initialize a new state machine that's ready for a new row.
+   *
+   * @param adapter The adapter that will build the final row.
+   */
+  StateMachine(RowBuilder<RowT> adapter) {
+    this.adapter = adapter;
+    reset();
+  }
+
+  /**
+   * Handle last scanned row events from the server.
+   *
+   * <p>The server might return the row key of the last row it has scanned. The client can use this
+   * to construct a more efficient retry request if needed: any row keys or portions of ranges less
+   * than this row key can be dropped from the request. The retry logic is implemented downstream
+   * from the logical rows produced by this class. To communicate this information to the retry
+   * logic, the state machine will encode the last scanned keys as scan marker rows using {@link
+   * RowBuilder#createScanMarkerRow(ByteString)}. It is the responsibility of the retry logic to
+   * filter out these marker before delivering the results to the user.
+   *
+   * <p>The scan marker must be immediately consumed before any more events are processed.
+   *
+   * <dl>
+   *   <dt>Valid states:
+   *   <dd>{@link StateMachine#AWAITING_NEW_ROW}
+   *   <dt>Resulting states:
+   *   <dd>{@link StateMachine#AWAITING_ROW_CONSUME}
+   * </dl>
+   */
+  void handleLastScannedRow(ByteString key) {
+    try {
+      currentState = currentState.handleLastScannedRow(key);
+    } catch (RuntimeException e) {
+      currentState = null;
+      throw e;
+    }
+  }
+
+  /**
+   * Feeds a new chunk into the sate machine. If the chunk is invalid, the state machine will throw
+   * an exception and should be used for further input.
```

Comment:
Is it "should not"? `currentState` being null seems to not play well with calling this method again in the future.

Reply:
fixed